### PR TITLE
implementation of Fletcher-Reeves Algorithm

### PR DESCRIPTION
**Pull request by ita9naiwa**
+ Polak-Ribiere Method;
To my knowledge, it was quite successful to use conjugate gradient variants on general nonconstrained optimization

This PR depends on Line Search of PR #128.

- Beta division is required to guarantee strong Wolfe Condition, but (i don't know) it raises error..

### CODE REVIEW

**Comment by mblondel** [comment]
Thanks for starting this! Once the line search module is merged, I agree that nonlinear CG could be nice to have in the library. Scipy seems to use the  Polak-Ribiere variant (see [this page](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.minimize.html)). I would therefore suggest to name the class `ConjugateGradient` and to have a flag `method="polak-ribiere"|fletcher-reeves"` to switch between the two. @shoyer and @fabianp may have some thoughts on this too.

**Comment by ita9naiwa** [comment]
@mblondel Thanks for replying!
I changed the classname to`ConjugateGradient`, and argname to `method` you suggested!
 
https://github.com/google/jaxopt/blob/ee976efa3c2abab26e2b94bab6793748bc036b5d/jaxopt/_src/nonlinear_cg.py#L156-L167

I think it's good to set filename to be `nonlinear_cg.py` 
since, there is [linear_cg](https://jax.readthedocs.io/en/latest/_autosummary/jax.scipy.sparse.linalg.cg.html) in scipy and jax already and might be confusing.

**Comment by mblondel** [comment]
Good point, I agree that `nonlinear_cg.py` will avoid confusions. In that case, it's probably better to call the class `NonlinearCG` too.

**Comment by ita9naiwa** [comment]
I changed the names of classes and arguments and I guess it's not confusing now.

I think this implementation works well yet I'm not sure it's fast (I'm quite new to jax and not sure I'm employing jax's functionality well)

**Comment by ita9naiwa** [comment]
and I made some commits while I'm at work and I want to amend it but it's not very easy to do with other's un-merged commits. (I couldn't do CLA yet)

I'll clean it up after Line Search PR will be merged..

**Comment by mblondel** [comment]
Thanks @ita9naiwa! I agree it's better to wait for the other PR to be merged.

**Comment by ita9naiwa** [comment]
@mblondel I cleaned up commits, and I think it seems ready for review now.

File: jaxopt/__init__.py [hunk 1]
```diff
@@ -31,6 +31,7 @@
 from jaxopt._src.gradient_descent import GradientDescent
 from jaxopt._src.iterative_refinement import IterativeRefinement
 from jaxopt._src.lbfgs import LBFGS
+from jaxopt._src.nonlinear_cg import NonlinearCG
```
Comment (mblondel) [inline jaxopt/__init__.py:34]: Move below mirror descent.

File: jaxopt/_src/nonlinear_cg.py [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+  stepsize: float
+  error: float
+  value: float
+  grad: jnp.ndarray
+  p: jnp.ndarray
+  aux: Optional[Any] = None
+
+
+@dataclass(eq=False)
+class NonlinearCG(base.IterativeSolver):
```
Comment (mblondel) [inline jaxopt/_src/nonlinear_cg.py:49]: Can you add the class to `docs/api.rst` and `docs/unconstrained.rst`?

**Review by mblondel — CHANGES_REQUESTED** [review]
Thanks a lot @ita9naiwa for this nice contribution! I left many detailed comments but overall this looks quite good, thanks!

File: jaxopt/_src/nonlinear_cg.py (hunk 2 above)
Comment (ita9naiwa) [reply jaxopt/_src/nonlinear_cg.py:49]: I added!

File: jaxopt/_src/nonlinear_cg.py [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+    new_params = tree_add_scalar_mul(params, new_stepsize, p)
+    (new_value, new_aux), new_grad = self._value_and_grad_with_aux(new_params, *args, **kwargs)
+
+    if self.method == "polak-ribiere":
+      new_beta = tree_div(tree_vdot(new_grad, tree_sub(new_grad, grad)), 1e-6 + tree_vdot(grad, grad))
+      new_beta = jnp.clip(new_beta, a_min=0)
+    elif self.method == "fletcher-reeves":
+      new_beta = tree_div(tree_vdot(new_grad, new_grad), 1e-6 + tree_vdot(grad, grad))
+    elif self.method == 'hestenes-stiefel':
+      new_beta = tree_div(tree_vdot(new_grad, tree_sub(new_grad, grad)), 1e-6 + tree_vdot(p, tree_sub(new_grad, grad)))
```
[the remark this answers is not in the document]
Comment (ita9naiwa) [reply jaxopt/_src/nonlinear_cg.py:158]: I copy-pasted `See Numerical Optimization, second edition, equation (5.#).`
three times, but not sure it's good..

File: jaxopt/_src/nonlinear_cg.py [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+  p: jnp.ndarray
+  aux: Optional[Any] = None
+
+
+@dataclass(eq=False)
+class NonlinearCG(base.IterativeSolver):
+  """
+    Nonlinear Conjugate Gradient Solver.
+  Attributes:
+    fun: a smooth function of the form ``fun(x, *args, **kwargs)``.
```
[the remark this answers is not in the document]
Comment (ita9naiwa) [reply jaxopt/_src/nonlinear_cg.py:53]: I couldn't find an appropriate example to follow about enumerating possible strings in docstrings. so I followed similar to scikit-learn's one.

**Comment by ita9naiwa** [comment]
@mblondel  Thanks for the review!

I fixed it and made some of your comments unresolved if I am not certain about my modification.

File: jaxopt/_src/nonlinear_cg.py [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+    method: which variant to calculate the beta parameter in Nonlinear CG.
+    ``{"polak-ribiere", "fletcher-reeves", "hestenes-stiefel"}``
```
Comment (mblondel) [inline jaxopt/_src/nonlinear_cg.py:55]:
```suggestion
    [two spaces]Possible values are `"polak-ribiere"`, `"fletcher-reeves"` and "`hestenes-stiefel"`.
```

Comment (mblondel) [reply jaxopt/_src/nonlinear_cg.py:55]: Make sure you add the two spaces at the beginning of the line, otherwise it may not be considered as the attribute description.

File: jaxopt/_src/nonlinear_cg.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (mblondel) [reply jaxopt/_src/nonlinear_cg.py:53]: I suggested an edit below. Marking this one as resolved.

File: jaxopt/_src/nonlinear_cg.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mblondel) [reply jaxopt/_src/nonlinear_cg.py:158]: Fine with me, thanks :)

File: jaxopt/_src/nonlinear_cg.py (hunk 5 above)
Comment (ita9naiwa) [reply jaxopt/_src/nonlinear_cg.py]: I did not know that. thanks to let me know. It looks fine to me now.

**Comment by mblondel** [comment]
Thanks @ita9naiwa! A few comments are not resolved yet.

**Comment by ita9naiwa** [comment]
@mblondel I missed the comments and fixed them and it works on the test well.

I think `1e-6` looks clumsy and can be moved as an argument of the class, how do you think?

**Comment by mblondel** [comment]
I don't think it's a good idea to expose `eps` as a class option. Maybe use this instead: `eps = jnp.finfo(params.dtype).eps`

**Comment by ita9naiwa** [comment]
I think `eps = jnp.finfo(params.dtype).eps` gives too small value (1e-16 for float), I think then it's okay to keep as is(use  constant number 1e-6) or define `eps=1e-6` and use it.

**Comment by mblondel** [comment]
In that case defining `eps = 1e-6` locally in the function and using it sounds good to me!

File: docs/unconstrained.rst [hunk 6]
```diff
@@ -52,7 +53,7 @@ instantiated and run as follows::
   # Alternatively, we could have used one of these solvers as well:
   # solver = jaxopt.GradientDescent(fun=ridge_reg_objective, maxiter=500)
   # solver = jaxopt.ScipyMinimize(fun=ridge_reg_objective, method="L-BFGS-B", maxiter=500)
-
+  # cg_model = NonlinearCG(fun=ridge_reg_objective, maxiter=300, method="polak-ribiere")
```
Comment (mblondel) [inline docs/unconstrained.rst:56]:
```suggestion
  # cg_model = jaxopt.NonlinearCG(fun=ridge_reg_objective, maxiter=300, method="polak-ribiere")
```

Comment (ita9naiwa) [reply docs/unconstrained.rst]: I'm sorry, I fixed it again.

**Comment by ita9naiwa** [comment]
for me, it seems to look pretty now. Thanks for your thorough and kind review and re-review and on
I really appreciate it! @mblondel

**Comment by mblondel** [comment]
Thanks a lot for your hard work! The only remaining thing is to squash your commits into a single commit. Then we'll merge!

**Comment by ita9naiwa** [comment]
I squashed up all commits into a single commit, and every test works fine. Thanks!

**Comment by mblondel** [comment]
Merged, thanks a lot!
